src/csharp/Microsoft.Spark/Utils/AssemblyLoader.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Spark.Utils
+{
+    internal static class AssemblyLoader
+    {
+        internal static Func<string, Assembly> LoadFromFile { get; set; } = Assembly.LoadFrom;
+
+        internal static Func<string, Assembly> LoadFromName { get; set; } = Assembly.Load;
+
+        private static readonly Dictionary<string, Assembly> s_assemblyCache =
+            new Dictionary<string, Assembly>();
+
+        private static readonly string[] s_searchPaths =
+            new[] { Directory.GetCurrentDirectory(), AppDomain.CurrentDomain.BaseDirectory };
+
+        private static readonly string[] s_extensions =
+            RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+                    new[] { ".dll", ".exe", ".ni.dll", ".ni.exe" } :
+                    new[] { ".dll", ".ni.dll" };
+
+        private static readonly object s_cacheLock = new object();
+
+        /// <summary>
+        /// Return the cached assembly, otherwise attempt to load and cache the assembly
+        /// in the following order:
+        /// 1) Load the assembly from disk using assemblyFileName.
+        /// 2) Search for assemblyName in the current app domain.
+        /// </summary>
+        /// <param name="assemblyName">The full name of the assembly</param>
+        /// <param name="assemblyFileName">Name of the file that contains the assembly</param>
+        /// <returns>Cached or Loaded Assembly</returns>
+        internal static Assembly LoadAssembly(string assemblyName, string assemblyFileName)
+        {
+            lock (s_cacheLock)
+            {
+                if (s_assemblyCache.TryGetValue(assemblyName, out Assembly assembly))
+                {
+                    return assembly;
+                }
+
+                if (!TryLoadAssembly(assemblyFileName, ref assembly))
+                {
+                    assembly = LoadFromName(assemblyName);
+                }
+
+                s_assemblyCache[assemblyName] = assembly;
+                return assembly;
+            }
+        }
+
+        /// <summary>
+        /// Return the cached assembly, otherwise look in the following probing paths,
+        /// searching for the simple assembly name and s_extension combination.
+        /// 1) The working directory
+        /// 2) The directory of the application
+        /// </summary>
```

Comment:
I think it would be beneficial to add a comment why we decided to go with this order (in the cluster run, the dlls are usually shipped to the working directory, for example) since this is very specific to our use case.

Reply:
Added some extra details regarding the probing path order.